hello-clusters notebook: Perform and evaluate clustering (#874)
AlexsLemonade/OpenScPCA-analysis: analyses/hello-clusters/01_perform-evaluate-clustering.Rmd

If your 
The function assumes the PC name is `"PCA"`. 
If your SCE has a different name for the principal components, you can provide the name with the argument `pc_name`.

We also strongly recommend providing a random seed with the `seed` argument to ensure reproducibility.

```{r cluster sce}
# Calculate clusters with default parameters
cluster_results <- calculate_clusters(
  sce,
  seed = params$seed
)

# Print resulting table
cluster_results
```

The following chunk shows an example of performing clustering with several non-default parameters:

```{r cluster sce nondefault}
# Calculate clusters with non-default parameters
cluster_results <- calculate_clusters(
  sce,
  algorithm = "leiden",
  nn = 15,
  objective_function = "modularity",
  seed = params$seed
)

# Print resulting table
cluster_results
```

### Clustering on a Seurat object

When performing clustering on a Seurat object, `rOpenScPCA::calculate_clusters` assumes that the principal components are stored in a reduced dimension slot named `"pca"`.

> If your Seurat has a different name for the principal components, you can provide the name with the argument `pc_name`.

Reviewer: ```suggestion
If your Seurat object has a different name for the principal components, you can provide the name with the argument `pc_name`.
```

In this example, because we created the Seurat object from the SCE object, its principal components are named `"PCA"`, so we will provide that information.

We also strongly recommend providing a random seed with the `seed` argument to ensure reproducibility.

```{r cluster seurat}
cluster_results <- calculate_clusters(
  seurat_obj,
  # Since we converted this from an SCE object, its PC name is `"PCA"`
  pc_name = "PCA",
  seed = params$seed
)

cluster_results
```

## Evaluate clustering results

This section demonstrates how to use several functions for evaluating cluster quality and reliability. 
All examples will use the SCE object for demonstration, but a Seurat object can be provided instead.

All functions presented in this section take the following required arguments:

> * An SCE or Seurat object that contains principal components

Reviewer: Can't we also just pass in the PCA matrix?

This value ranges from 0 to 1.
Cells in well-separated clusters should have high purity values closer to 1, since there should be minimal overlap between member and neighboring cells.
You can read more about neighborhood purity from the [_Orchestrating Single Cell Analysis with Bioconductor_ book](https://bioconductor.org/books/3.19/OSCA.advanced/clustering-redux.html#cluster-purity).

We'll use the function `rOpenScPCA::calculate_purity()` to calculate the neighborhood purity.

This function will return the inputted data frame with two additional columns:

* `purity`: The neighborhood purity for the cell
* `maximum_neighbor`: The cluster with the highest proportion of observations neighboring the cell

```{r purity}
# calculate the neighborhood purity for each cell
purity_results <- rOpenScPCA::calculate_purity(
  pca_matrix,
  cluster_results_df
)

# Print the first rows of the resulting table
head(purity_results)
```

We can visualize these results by plotting purity clusters as violin plots, for example:

```{r violin purity}
ggplot(purity_results) +
  aes(x = cluster, y = purity) +

>   geom_violin(fill = "darkolivegreen3") +

Author: > I do care that when you are plotting different metrics you should use different colors.

Now _that_ I can do 🎨

```

In this example, we'll show how to use the cluster evaluation functions on these clusters.

To begin, we'll prepare a data frame with two columns: `cell_id` containing cell barcodes, and `cluster` containing the cluster identities.

```{r prepare scpca data frame}
scpca_cluster_df <- data.frame(
  cell_id = colnames(sce),
  cluster = sce$cluster
)

head(scpca_cluster_df)
```

We can run evaluation functions using this data frame and the SCE object.

```{r scpca silhouette}
# Calculate silhouette width
scpca_silhouette_df <- rOpenScPCA::calculate_silhouette(
  sce,
  scpca_cluster_df
)
```

```{r scpca purity}
# Calculate neighborhood purity
scpca_purity_df <- rOpenScPCA::calculate_purity(
  sce,
  scpca_cluster_df
)
```

When running `rOpenScPCA::calculate_stability()`, we'll specify the same parameters originally used to build the clusters.
Generally speaking, we only recommend evaluating clusters with `rOpenScPCA::calculate_stability()` if you know the original parameters used.

```{r scpca stability}
scpca_stability_df <- rOpenScPCA::calculate_stability(
  sce,
  scpca_cluster_df,
  # provide ScPCA clustering parameters
  algorithm = "louvain",
  weighting = "jaccard",

>   nn = 20

Author: I feel fairly sure the answer here isn't "let's add another dependency!", but I did think this was useful to know about https://cran.r-project.org/web/packages/strex/vignettes/argument-matching.html

module_base <- file.path(repository_base, "analyses", "hello-clusters")
```

### Input file

```{r paths}
# Path to processed SCE file for sample SCPCS000001
input_sce_file <- file.path(data_dir, "SCPCP000001", "SCPCS000001", "SCPCL000001_processed.rds")
```

## Read in and prepare data

To begin, we'll read in the `SingleCellExperiment` (SCE) object.
We'll also establish a corresponding processed Seurat object from its raw counts that we'll use for some examples.

```{r read data}
# Read the SCE file
sce <- readRDS(input_sce_file)
```

```{r create process seurat, message = FALSE}
# Convert to a Seurat object
seurat_obj <- CreateSeuratObject(counts = counts(sce), assay = "RNA")

> 

Author: This needs to be updated with `rOpenScPCA::sce_to_seurat()` once https://github.com/AlexsLemonade/rOpenScPCA/pull/15 goes in.

# Process the object with a Seurat pipeline to obtain clusters
seurat_obj <- seurat_obj |>
  SCTransform() |>
  RunPCA() |>
  FindNeighbors() |>
  FindClusters()
```

> We'll also extract its PCA matrix, which we'll use to demonstrate how to calculate and evaluate clusters.

Author: ```suggestion
For the initial cluster calculations and evaluations, we will use the PCA matrix extracted from the SCE object.
It's also possible to use an SCE object or a Seurat object directly, which we will demonstrate later.
```

This function will return a data frame with columns `replicate`, `ari`, and additional columns for the clustering parameters used when calculating bootstrap clusters.

```{r stability, warning=FALSE}
# calculate the stability of clusters
stability_results <- calculate_stability(
  pca_matrix,
  cluster_results_df
)

# print the result
stability_results
```

#### Using non-default clustering parameters

When calculating bootstrap clusters, `rOpenScPCA::calculate_stability()` uses `rOpenScPCA::calculate_clusters()` with default parameters. 
If your original clusters were not calculated with these defaults, you should pass those customized values into this function as well to ensure a fair comparison between your original clusters and the bootstrap clusters.

```{r stability custom parameters}
# Calculate clusters with non-default parameters
cluster_df_leiden <- calculate_clusters(
  pca_matrix,
  algorithm = "leiden",
  resolution = 0.5,
  nn = 15
)

# Now, pass in the same arguments customizing parameters here
stability_results_leiden <- calculate_stability(
  pca_matrix,
  cluster_df_leiden,
  algorithm = "leiden",
  resolution = 0.5,
  nn = 15
)

stability_results_leiden
```

> ##  Calculate QC metrics on existing clusters

Reviewer: I would move this to after the object section and demonstrate these using the objects themselves.

> Even if you perform clustering with other tools, you can still use the `rOpenScPCA` functions to evaluate your clusters.

Reviewer: ```suggestion
If you already have clustering results calculated with other tools, you can still use the `rOpenScPCA` functions to evaluate your clusters.
```

In this section, we'll present examples of how you can calculate the silhouette width, neighborhood purity, and cluster stability from existing cluster assignments.

### Evaluating Seurat clusters

If you are analyzing your data with a Seurat pipeline that includes calculating clusters, you can use `rOpenScPCA` to evaluate them.

> We previously created a Seurat object called `seurat_obj` with clusters:

Reviewer: Create the object here, I think.

```{r show seurat clusters}
head(seurat_obj$seurat_clusters)
```

To calculate QC metrics on these clusters, we'll need two pieces of information:

- The PCA matrix
- A data frame with columns `cell_id` and `cluster`

We'll create those here:

```{r prepare seurat input}
# Extract the PCA matrix from the Seurat object
seurat_pca_matrix <- Embeddings(seurat_obj, reduction = "pca")

# Create a data frame for input
seurat_cluster_df <- data.frame(
  cell_id = names(seurat_obj$seurat_clusters),
  cluster = seurat_obj$seurat_clusters
)
```

Now, we can run `calculate_silhouette()` and `calculate_purity()`:

```{r seurat silhouette}
seurat_silhouette_df <- calculate_silhouette(
  seurat_pca_matrix,
  seurat_cluster_df
)
seurat_silhouette_df
```

```{r seurat purity}
seurat_purity_df <- calculate_purity(
  seurat_pca_matrix,
  seurat_cluster_df
)
seurat_purity_df
```

> We do not recommend using `rOpenScPCA::calculate_stability()` on Seurat clusters due to differences in the underlying clustering approach between Seurat and the `bluster` package which `rOpenScPCA` uses.

Reviewer: 👍🏼 I was all set from the intro to this part to say you needed a note about this.

### Evaluating ScPCA clusters

ScPCA cell metadata already contains a column called `cluster` with results from an automated clustering. 
These clusters were calculated using `bluster`, the same tool that `rOpenScPCA` uses, with the following parameters:

* The Louvain algorithm
* Jaccard weighting

> * 20 nearest neighbors

Reviewer: I would actually show that we have these parameters stored in the object, and use those to retrieve the values:

```
metadata(sce)$cluster_algorithm
metadata(sce)$cluster_weighting
metadata(sce)$cluster_nn
```

In this example, we'll show how to use the cluster evaluation functions on these clusters.

To begin, we'll prepare a data frame with two columns: `cell_id` containing cell barcodes, and `cluster` containing the cluster identities.

```{r prepare scpca data frame}
scpca_cluster_df <- data.frame(
  cell_id = colnames(sce),

>   cluster = colData(sce)$cluster

Reviewer: SCE objects are smart; no need for `colData` here!
```suggestion
  cluster = sce$cluster
```

Instead of a matrix, you can alternatively pass in an SCE or Seurat object that contains a matrix.
We show examples of this below with `rOpenScPCA::calculate_clusters()`, but this will also work for any of the evaluation functions as well:

```{r run on sce}
# Calculate clusters from an SCE object using default parameters
cluster_results_df <- calculate_clusters(sce)
cluster_results_df
```

```{r run on seurat}
# Calculate clusters from a Seurat object using default parameters
cluster_results_df <- calculate_clusters(seurat_obj)
cluster_results_df
```

`rOpenScPCA` assumes that the PCA matrix is named `PCA` in SCE objects, and `pca` in Seurat objects (as they were in the above examples).
If the PCA matrix you want to use in the object has a different name, you can provide the argument `pc_name`.
We can see this below with an SCE object, for example:

```{r use pc_name}
# First, we'll rename the PCA matrix for demonstration
reducedDimNames(sce) <- c("PCA_matrix", "UMAP")
reducedDimNames(sce)

# Calculate clusters from an SCE object using default parameters
cluster_results_df <- calculate_clusters(
  sce,
  pc_name = "PCA_matrix"
)
cluster_results_df

> ```

Reviewer: This feels like a detail we don't need to demonstrate. I would remove this section, or at least the code for it.

## Saving clustering results 

Results can either be directly exported as a TSV file (e.g., with `readr::write_tsv()`), or you can add the results into your SCE or Seurat object.
The subsequent examples will show you how to save the `cluster_results_df` data frame to an SCE and a Seurat object.

_A word of caution!_
Objects from the ScPCA Portal already contain a column called `cluster` with results from an automated clustering.
These automatic clusters were not evaluated, and their parameters were not optimized for any given library.

To avoid ambiguity between the existing and new clustering results, you may want to rename incoming columns.

We'll therefore begin by adding the prefix `ropenscpca_` to all column names in the `cluster_results_df` data frame before adding them to a given object.

```{r rename columns}
# First, rename columns in `cluster_results_df` to avoid ambiguity
cluster_results_renamed_df <- cluster_results_df |>

>   # add the prefix openscpca_ to all columns

Reviewer: ```suggestion
  # add the prefix "ropenscpca_" to all columns
```

>   dplyr::rename_with(~ paste0("ropenscpca_", .x))

Reviewer: Modern R
```suggestion
  dplyr::rename_with(\(x) paste0("ropenscpca_", x))
```

cluster_results_renamed_df
```

### Saving results to an SCE object 

Follow these steps to save the clustering results to an SCE object's `colData`:

```{r add to sce}
# First, extract the colData into a separate data frame for manipulation
sce_coldata <- colData(sce) |>
  as.data.frame()

# Second, add columns as desired using the barcodes and cell_id columns to join
sce_coldata <- sce_coldata |>
  dplyr::left_join(
    cluster_results_renamed_df,
    by = c("barcodes" = "ropenscpca_cell_id")
  )

# Third, add the modified colData back into the SCE
# Use the row.names argument to ensure rownames (barcodes) are retained; don't forget this step!
colData(sce) <- DataFrame(
  sce_coldata,
  row.names = sce_coldata$barcodes
)

colData(sce)
```

### Saving results to a Seurat object 

Follow these steps save the clustering results to a Seurat object's cell metadata:

```{r add to seurat}
# Add the cluster result data frame to the Seurat object metadata

> seurat_obj <- AddMetaData(seurat_obj, cluster_results_renamed_df)

Reviewer: Does Seurat assume that the data frame is in the same order? If so, we might want to caution about that. 

In the interests of simplicity, you might also want to show making that assumption above, in which case you can first check that it is true with `all(colnames(sce) == cluster_results_df$cell_id)` or something like that, then just add the clusers more simply with `scr$ropenscpca_clusters <- cluster_results_df$cluster`. I don't think we really need to show adding all of the cluster parameters?
